[`Docs` / `BetterTransformer` ] Added more details about flash attention + SDPA (#25265)
huggingface/transformers: docs/source/en/perf_infer_gpu_many.md

model = AutoModelForCausalLM.from_pretrained("facebook/opt-350m").to("cuda")
# convert the model to BetterTransformer
model.to_bettertransformer()

input_text = "Hello my dog is cute and"
inputs = tokenizer(input_text, return_tensors="pt").to("cuda")

with torch.backends.cuda.sdp_kernel(enable_flash=True, enable_math=False, enable_mem_efficient=False):
    outputs = model.generate(**inputs)

print(tokenizer.decode(outputs[0], skip_special_tokens=True))
```

If you see a bug with a traceback saying 

```bash
RuntimeError: No available kernel.  Aborting execution.
```

you may try to use the PyTorch nightly version, which may have a larger coverage for Flash Attention:

```bash
pip3 install -U --pre torch torchvision torchaudio --index-url https://download.pytorch.org/whl/nightly/cu118

Have a look at this [blog post](https://pytorch.org/blog/out-of-the-box-acceleration/) to learn more about what is possible to do with the `BetterTransformer` + SDPA API.

### Encoder models

For encoder models the API will dispatch the forward call of encoder layers to `torch.nn.TransformerEncoder` that will execute the fast path implementation of the encoder layers.

> Have a look at this [blog post](https://medium.com/pytorch/bettertransformer-out-of-the-box-performance-for-huggingface-transformers-3fbe27d50ab2) and [this one from PyTorch](torch.nn.TransformerEncoder) to learn more about what is possible to do with `BetterTransformer` API for encoder models.

Reviewer: ```suggestion
More details can be found in [this blog post](https://medium.com/pytorch/bettertransformer-out-of-the-box-performance-for-huggingface-transformers-3fbe27d50ab2) and [this second one](https://pytorch.org/blog/a-better-transformer-for-fast-transformer-encoder-inference/).
```

</Tip>

## BetterTransformer

> [`BetterTransformer`](https://huggingface.co/docs/optimum/bettertransformer/overview) API converts 🤗 transformers models to make them use PyTorch-native transformer fastpath that calls optimized kernels such as Flash Attention under the hood.  

Reviewer 2: ```suggestion
[BetterTransformer](https://huggingface.co/docs/optimum/bettertransformer/overview) converts 🤗 Transformers models to use the PyTorch-native fastpath execution, which calls optimized kernels like Flash Attention under the hood.  
```

BetterTransformer is also supported for faster inference on multi-GPU for text, image, and audio models.

<Tip>

Flash Attention can only be used for models using fp16 or bf16 dtype. Make sure to cast your model before using BetterTransformer.
  
</Tip>

### Decoder models

> For text models, especially decoder-based models (GPT, T5, Llama, etc.), the `BetterTransformer` API converts all attention operations to use the [`torch.nn.functional.scaled_dot_product_attention` operator](https://pytorch.org/docs/master/generated/torch.nn.functional.scaled_dot_product_attention) (SDPA) that is only available in PyTorch 2.0 and onwards. 

Reviewer 2: ```suggestion
For text models, especially decoder-based models (GPT, T5, Llama, etc.), the BetterTransformer API converts all attention operations to use the [`torch.nn.functional.scaled_dot_product_attention` operator](https://pytorch.org/docs/master/generated/torch.nn.functional.scaled_dot_product_attention) (SDPA) that is only available in PyTorch 2.0 and onwards. 
```

> To convert a model to `BetterTransformer`:

Reviewer 2: ```suggestion
To convert a model to BetterTransformer:
```

```python
from transformers import AutoModelForCausalLM

model = AutoModelForCausalLM.from_pretrained("facebook/opt-350m")
# convert the model to BetterTransformer
model.to_bettertransformer()

# Use it for training or inference
```

> SDPA can also call [Flash-Attention](https://arxiv.org/abs/2205.14135) kernels under the hood. In order to force the usage of Flash Attention or check that Flash Attention is available in a given setting (hardware, problem size), the context manager [`with torch.backends.cuda.sdp_kernel(enable_flash=True, enable_math=False, enable_mem_efficient=False)`](https://pytorch.org/docs/master/backends.html#torch.backends.cuda.sdp_kernel) can be used:

Reviewer 2: I think this makes it a bit easier to read, and then we can highlight the relevant line of code below to further emphasize it

```suggestion
SDPA can also call [Flash Attention](https://arxiv.org/abs/2205.14135) kernels under the hood. To enable Flash Attention or to check that it is available in a given setting (hardware, problem size), use [`torch.backends.cuda.sdp_kernel`](https://pytorch.org/docs/master/backends.html#torch.backends.cuda.sdp_kernel) as a context manager:
```

> ```python

Reviewer 2: ```suggestion
```diff
```

Author: nice!

import torch
from transformers import AutoModelForCausalLM, AutoTokenizer

tokenizer = AutoTokenizer.from_pretrained("facebook/opt-350m")
model = AutoModelForCausalLM.from_pretrained("facebook/opt-350m").to("cuda")
# convert the model to BetterTransformer
model.to_bettertransformer()

input_text = "Hello my dog is cute and"
inputs = tokenizer(input_text, return_tensors="pt").to("cuda")

> with torch.backends.cuda.sdp_kernel(enable_flash=True, enable_math=False, enable_mem_efficient=False):
>     outputs = model.generate(**inputs)

Reviewer 2: ```suggestion
+ with torch.backends.cuda.sdp_kernel(enable_flash=True, enable_math=False, enable_mem_efficient=False):
    outputs = model.generate(**inputs)
```

print(tokenizer.decode(outputs[0], skip_special_tokens=True))
```

If you see a bug with a traceback saying 

```bash
RuntimeError: No available kernel.  Aborting execution.
```

> you may try to use the PyTorch nightly version, which may have a larger coverage for Flash Attention:

Reviewer 2: ```suggestion
try using the PyTorch nightly version, which may have a broader coverage for Flash Attention:
```

```bash

> pip3 install -U --pre torch torchvision torchaudio --index-url https://download.pytorch.org/whl/nightly/cu118

Reviewer 2: Close the code snippet here :)

Author: Done!

> Have a look at this [blog post](https://pytorch.org/blog/out-of-the-box-acceleration/) to learn more about what is possible to do with the `BetterTransformer` + SDPA API.

Reviewer 2: ```suggestion
Have a look at this [blog post](https://pytorch.org/blog/out-of-the-box-acceleration/) to learn more about what is possible with the BetterTransformer + SDPA API.
```

### Encoder models

> For encoder models during inference, BetterTransformer will dispatch the forward call of encoder layers to an equivalent of [`torch.nn.TransformerEncoderLayer`](https://pytorch.org/docs/stable/generated/torch.nn.TransformerEncoderLayer.html) that will execute the fast path implementation of the encoder layers.

Reviewer 2: ```suggestion
For encoder models during inference, BetterTransformer dispatches the forward call of encoder layers to an equivalent of [`torch.nn.TransformerEncoderLayer`](https://pytorch.org/docs/stable/generated/torch.nn.TransformerEncoderLayer.html) that will execute the fastpath implementation of the encoder layers.
```

> As `torch.nn.TransformerEncoderLayer` fastpath does not support training, we instead dispatch to `torch.nn.functional.scaled_dot_product_attention` during training, that do not leverage nested tensors but can leverage Flash Attention or Memory-Efficient Attention fused kernels.

Reviewer 2: ```suggestion
Because `torch.nn.TransformerEncoderLayer` fastpath does not support training, it is dispatched to `torch.nn.functional.scaled_dot_product_attention` instead, which does not leverage nested tensors but can use Flash Attention or Memory-Efficient Attention fused kernels.
```

> More details can be found in [this blog post](https://medium.com/pytorch/bettertransformer-out-of-the-box-performance-for-huggingface-transformers-3fbe27d50ab2) and [this second one](https://pytorch.org/blog/a-better-transformer-for-fast-transformer-encoder-inference/).

Reviewer 2: ```suggestion
More details about BetterTransformer performance can be found in this [blog post](https://medium.com/pytorch/bettertransformer-out-of-the-box-performance-for-huggingface-transformers-3fbe27d50ab2), and you can learn more about BetterTransformer for encoder models in this [blog](https://pytorch.org/blog/a-better-transformer-for-fast-transformer-encoder-inference/).
```

</Tip>

> ## `BetterTransformer` API and Flash Attention for faster inference

Reviewer 2: The backticks aren't supported/rendered in the right navbar

```suggestion
## BetterTransformer and Flash Attention
```

> We have recently integrated `BetterTransformer` for faster inference on multi-GPU for text, image and audio models. Check the documentation about this integration [here](https://huggingface.co/docs/optimum/bettertransformer/overview) for more details. 

Reviewer 2: ```suggestion
[`BetterTransformer`](https://huggingface.co/docs/optimum/bettertransformer/overview) is supported for faster inference on multi-GPU for text, image, and audio models.
```

> For text models, especially decoder-based models (e.g. GPT, T5, Llama, etc.), the `BetterTransformer` API will convert all attention operations to use the [`torch.nn.functional.scaled_dot_product_attention` method](https://pytorch.org/docs/master/generated/torch.nn.functional.scaled_dot_product_attention) (SDPA), that is available only from PyTorch 2.0 and onwards. 

Reviewer 2: ```suggestion
For text models, especially decoder-based models (GPT, T5, Llama, etc.), the `BetterTransformer` API converts all attention operations to use the [`torch.nn.functional.scaled_dot_product_attention` method](https://pytorch.org/docs/master/generated/torch.nn.functional.scaled_dot_product_attention) (SDPA) that is only available in PyTorch 2.0 and onwards. 
```

> An example usage of the `BetterTransformer` API is shown below:

Reviewer 2: ```suggestion
To convert a model to `BetterTransformer`:
```

```python
from transformers import AutoModelForCausalLM

model = AutoModelForCausalLM.from_pretrained("facebook/opt-350m")
# convert the model to BetterTransformer
model.to_bettertransformer()

# Use it for training or inference
```

> According to the official documentation, `torch.nn.functional.scaled_dot_product_attention` can also call [Flash-Attention](https://arxiv.org/abs/2205.14135) kernels under the hood. If you want to force the usage of Flash Attention, you can use the [`torch.backends.cuda.sdp_kernel(enable_flash=True)`](https://pytorch.org/docs/master/backends.html#torch.backends.cuda.sdp_kernel) as below:

Reviewer 2: ```suggestion
SDPA can also call [Flash-Attention](https://arxiv.org/abs/2205.14135) kernels under the hood. If you want to force the usage of Flash Attention, use [`torch.backends.cuda.sdp_kernel(enable_flash=True)`](https://pytorch.org/docs/master/backends.html#torch.backends.cuda.sdp_kernel):
```

    outputs = model.generate(**inputs)

print(tokenizer.decode(outputs[0], skip_special_tokens=True))
```

If you see a bug with a traceback saying 

```bash
RuntimeError: No available kernel.  Aborting execution.
```

Install the PyTorch nightly version

```bash
pip3 install -U --pre torch torchvision torchaudio --index-url https://download.pytorch.org/whl/nightly/cu118
```

Or alternatively try to add an autocast context manager in addition to the SDPA context manager:

```python
import torch
from transformers import AutoModelForCausalLM, AutoTokenizer

tokenizer = AutoTokenizer.from_pretrained("facebook/opt-350m")
model = AutoModelForCausalLM.from_pretrained("facebook/opt-350m").to("cuda")
# convert the model to BetterTransformer
model.to_bettertransformer()

input_text = "Hello my dog is cute and"
inputs = tokenizer(input_text, return_tensors="pt").to("cuda")

with torch.cuda.amp.autocast(), torch.backends.cuda.sdp_kernel(enable_flash=True, enable_math=False, enable_mem_efficient=False):
    outputs = model.generate(**inputs)

print(tokenizer.decode(outputs[0], skip_special_tokens=True))
```

> For encoder models, the method [`~PreTrainedModel.reverse_bettertransformer`] allows to go back to the original modeling, which should be used before saving the model in order to use the canonical transformers modeling:

Reviewer 2: ```suggestion
For encoder models, the [`~PreTrainedModel.reverse_bettertransformer`] method reverts to the original model, which should be used before saving the model to use the canonical transformers modeling:
```

```python
model = model.reverse_bettertransformer()
model.save_pretrained("saved_model")
```

> Have a look at [this detailed blogpost](https://pytorch.org/blog/out-of-the-box-acceleration/) to read more about what is possible to do with `BetterTransformer` + SDPA API.

Reviewer 2: ```suggestion
Have a look at this [blog post](https://pytorch.org/blog/out-of-the-box-acceleration/) to learn more about what is possible to with the `BetterTransformer` + SDPA API.
```

</Tip>

## BetterTransformer

[`BetterTransformer`](https://huggingface.co/docs/optimum/bettertransformer/overview) API converts 🤗 transformers models to make them use PyTorch-native transformer fastpath that calls optimized kernels such as Flash Attention under the hood.  

BetterTransformer is also supported for faster inference on multi-GPU for text, image, and audio models.

> 

Reviewer: ```suggestion

<Tip>

Flash Attention can only be used for models using fp16 or bf16 dtype. Make sure to cast you model before using BetterTransformer.
  
</Tip>

```

### Decoder models

> For text models, especially decoder-based models (GPT, T5, Llama, etc.), the `BetterTransformer` API converts all attention operations to use the [`torch.nn.functional.scaled_dot_product_attention` method](https://pytorch.org/docs/master/generated/torch.nn.functional.scaled_dot_product_attention) (SDPA) that is only available in PyTorch 2.0 and onwards. 

Reviewer: ```suggestion
For text models, especially decoder-based models (GPT, T5, Llama, etc.), the `BetterTransformer` API converts all attention operations to use the [`torch.nn.functional.scaled_dot_product_attention` operator](https://pytorch.org/docs/master/generated/torch.nn.functional.scaled_dot_product_attention) (SDPA) that is only available in PyTorch 2.0 and onwards. 
```

To convert a model to `BetterTransformer`:

```python
from transformers import AutoModelForCausalLM

model = AutoModelForCausalLM.from_pretrained("facebook/opt-350m")
# convert the model to BetterTransformer
model.to_bettertransformer()

# Use it for training or inference
```

> SDPA can also call [Flash-Attention](https://arxiv.org/abs/2205.14135) kernels under the hood. If you want to force the usage of Flash Attention, use [`torch.backends.cuda.sdp_kernel(enable_flash=True)`](https://pytorch.org/docs/master/backends.html#torch.backends.cuda.sdp_kernel):

Reviewer: ```suggestion
SDPA can also call [Flash-Attention](https://arxiv.org/abs/2205.14135) kernels under the hood. If you want to force the usage of Flash Attention or check that Flash Attention is available in your setting (hardware, problem size), use [`torch.backends.cuda.sdp_kernel(enable_flash=True)`](https://pytorch.org/docs/master/backends.html#torch.backends.cuda.sdp_kernel):
```

> SDPA can also call [Flash-Attention](https://arxiv.org/abs/2205.14135) kernels under the hood. If you want to force the usage of Flash Attention, use [`torch.backends.cuda.sdp_kernel(enable_flash=True)`](https://pytorch.org/docs/master/backends.html#torch.backends.cuda.sdp_kernel):

Reviewer: `torch.backends.cuda.sdp_kernel(enable_flash=True)` is not enough. You need `torch.backends.cuda.sdp_kernel(enable_flash=True, enable_math=False, enable_mem_efficient=False` as below

```python
import torch
from transformers import AutoModelForCausalLM, AutoTokenizer

tokenizer = AutoTokenizer.from_pretrained("facebook/opt-350m")
model = AutoModelForCausalLM.from_pretrained("facebook/opt-350m").to("cuda")
# convert the model to BetterTransformer
model.to_bettertransformer()

input_text = "Hello my dog is cute and"
inputs = tokenizer(input_text, return_tensors="pt").to("cuda")

with torch.backends.cuda.sdp_kernel(enable_flash=True, enable_math=False, enable_mem_efficient=False):
    outputs = model.generate(**inputs)

print(tokenizer.decode(outputs[0], skip_special_tokens=True))
```

If you see a bug with a traceback saying 

```bash
RuntimeError: No available kernel.  Aborting execution.
```

> Install the PyTorch nightly version
> 
> ```bash
> pip3 install -U --pre torch torchvision torchaudio --index-url https://download.pytorch.org/whl/nightly/cu118
> ```

Reviewer: ```suggestion
you may try to use the PyTorch nightly version, which may have a larger coverage for Flash Attention:

```bash
pip3 install -U --pre torch torchvision torchaudio --index-url https://download.pytorch.org/whl/nightly/cu118
```
```

> Or alternatively try to add an autocast context manager in addition to the SDPA context manager:
> 
> ```python
> import torch
> from transformers import AutoModelForCausalLM, AutoTokenizer
> 
> tokenizer = AutoTokenizer.from_pretrained("facebook/opt-350m")
> model = AutoModelForCausalLM.from_pretrained("facebook/opt-350m").to("cuda")
> # convert the model to BetterTransformer
> model.to_bettertransformer()
> 
> input_text = "Hello my dog is cute and"
> inputs = tokenizer(input_text, return_tensors="pt").to("cuda")
> 
> with torch.cuda.amp.autocast(), torch.backends.cuda.sdp_kernel(enable_flash=True, enable_math=False, enable_mem_efficient=False):
>     outputs = model.generate(**inputs)
> 
> print(tokenizer.decode(outputs[0], skip_special_tokens=True))
> ```

Reviewer: ```suggestion
```

Have a look at this [blog post](https://pytorch.org/blog/out-of-the-box-acceleration/) to learn more about what is possible to do with the `BetterTransformer` + SDPA API.

### Encoder models

For encoder models, the [`~PreTrainedModel.reverse_bettertransformer`] method reverts to the original model, which should be used before saving the model to use the canonical transformers modeling:

```python
model = model.reverse_bettertransformer()
model.save_pretrained("saved_model")

> ```

Reviewer: I think we should not make the distinction between encoder / decoder models when it come to using `reverse_bettertransformer`.

For example, for encoder-decoder models (e.g. t5), both SDPA (in the decoder) and nestedtensor (in the encoder) are used. So in case one wants to save the model, he'll need to use `reverse_bettertransformer`.

To me the distinction is more in that you can get speedups for inference with encoder models (since nestedtensor is used), but for decoder models the speedup / dispatch to flash will only come (in pytorch 2.0) for training & batch size = 1 for inference.

Author: Thanks for the suggestion! I refactored a bit that section and removed the `reverse_bettertransformer` part as it is relevant only for training (that section is for inference only)
